system_metrics_collector/src/topic_statistics_collector/received_message_age.hpp

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TOPIC_STATISTICS_COLLECTOR__RECEIVED_MESSAGE_AGE_HPP_
+#define TOPIC_STATISTICS_COLLECTOR__RECEIVED_MESSAGE_AGE_HPP_
+
+#include <chrono>
+#include <string>
+
+#include "topic_statistics_collector.hpp"
+
+#include "message_filters/message_traits.h"
+#include "rcl/time.h"
+#include "rclcpp/clock.hpp"
+
+
+namespace topic_statistics_collector
+{
+/**
+ * Class used to measure the received messsage, tparam T, age from a ROS2 subscriber.
+ *
+ * @tparam T the message type to receive from the subscriber / listener
+*/
+template<typename T>
+class ReceivedMessageAgeCollector : public TopicStatisticsCollector<T>
+{
+public:
+  explicit ReceivedMessageAgeCollector(
+    const rclcpp::Clock & clock = rclcpp::Clock{RCL_STEADY_TIME})
```

Comment:
Is it possible that the message timestamp can originate from another clock type?

Reply:
Yes, we cannot control what type of `rcl_clock_type_t` incoming messages will have. In order to avoid  throwing from this method due to mismatched clock types, I'll make the following changes:

1. Make the source configurable in this ctor so subscribers can set it externally
2. Compare the clock sources in `OnMessageReceived`, print a `WARNING` if they're not equal and don't calculate message age